CI tests with github-actions and some improvements to testing

mezotaken opened the pull request:
**Describe what this pull request is trying to achieve.**

Use github-actions workflow to test every pull request automatically. Fix and improve tests overall. 

**Additional notes and description of your changes**
Changes:
1) Added empty checkpoint in test_files dir to be used in tests by default if no other checkpoint is specified
2) Server polling now checks whether the sdwebui process is still on and will cancel polling immediately on crash instead of on timeout
3) DDIM added to vanilla sampler txt2img test
4) options_write utility test ~~temporarily disabled~~ removed because the functionality is broken at the moment.
5) Typo fixed in txt2img with tiling test name
6) Tests separated in two folders. basic_features tests require no additional downloads, they could be run on cpu quite quickly. advanced_features tests should check **correctness** of results, so they require clip model, sd model, interrogate, upscale, face restore models etc. Some of them will require cuda device too (probably). 
7) Tests are run with --tests argument as before, but now it is possible to specify the directory: [--tests advanced_features]. If no options were provided then all tests will be ran.
8) Autotesting with github-actions will use basic_features tests for now, output of the main program is available as an uploaded artifact under job summary:
![image](https://user-images.githubusercontent.com/32306715/201717742-949a0d7a-701d-48fd-be2c-48dd0cc9c9bb.png)

As i stated, the goal is to run autotesting on every PR/commit, so it will be on CPU. 
Currently it is possible to run tests on cpu like that and it's quite fast:
`python launch.py --tests basic_features --no-half --disable-opt-split-attention --use-cpu all --skip-torch-cuda-test`
Face restoration cant be done on cpu for some reason, did not investigate it further. Probably FR model is loaded on cuda device no matter the --use-cpu option.

**Environment this was tested in**

 - OS: Windows 10
 - Browser: Chrome
 - Graphics card: NVIDIA GTX 1060 6GB

mezotaken committed:
Tests cleaned up

mezotaken committed:
Tests separated for github-actions CI

mezotaken committed:
skip cuda test

mezotaken commented:
Well. Looks like localhost connections are refused withing github-actions without some trickery. Also i need to propagate failure exitcode to the main program.

mezotaken committed:
Propagate test error and try it without localhost

mezotaken committed:
Use localhost with 80 port, count errors as well

mezotaken committed:
Use 80 port on launch

mezotaken committed:
Stop exporting cl args and upload stdout and stderr as artifacts

mezotaken committed:
remove test requiring codeformers

mezotaken committed:
Use empty model as CLIP weights

salco commented on test/basic_features/utils_test.py:
```diff
@@ -18,19 +18,19 @@ def setUp(self):
-  def test_options_write(self):
-    response = requests.get(self.url_options)
-    self.assertEqual(response.status_code, 200)
+  # def test_options_write(self):
```
Is it temporary and you will fix it before the merge or just dead code?
If it's just unused you should remove it completely. If someone need it later he can look at the git commit.

mezotaken replied:
Sure, i can remove it altogether. I'm not so experienced in the group workflow, so preferred to left it like this for a while rather than deleting it.

salco replied:
I will let the veto on @AUTOMATIC1111  but in my opinion it should be remove before the merge.

mezotaken replied:
Oh i have no objections to the removal. i just wanted to do it along with coverage. But there was some problems with adding it, so i'll just remove this test and fix some more issues. I've tried it locally https://github.com/mezotaken/stable-diffusion-webui/pull/2 but very unsuccesful. While using API i managed to succesfully run and show result, but due to testing being made through REST API between two independent processes it just measures tests. 
With different approach, when i use `coverage run`  on subprocess with webui instance it cannot generate coverage data file, tho locally it works. But its's wrong anyway 
![image](https://user-images.githubusercontent.com/32306715/202847825-4e104bac-c0f3-489a-8867-1cfa6643cb48.png)
![image](https://user-images.githubusercontent.com/32306715/202847833-c9904a68-01d3-467f-b064-651540df3862.png)
I dont get it, maybe will try again later in different PR

salco replied:
Yes no need to stale this PR. Let's add that in another PR. 👍

salco reviewed (commented):
There is some test commented that will be push is it what you want?

salco commented:
Also @mezotaken,  not sure how easy it is to add it but i though that it might be fun to also add code coverage to the report.
On older projet and with Jenkins i used this flow:
```
sh "pip install coverage"
sh "python -m coverage run -m unittest -v"
sh "python -m coverage xml"

publishCoverage adapters: [cobertura('tests/unittest/coverage.xml')], calculateDiffForChangeRequests: true, sourceFileResolver: sourceFiles('STORE_LAST_BUILD')
```

Link on [coverage.py](https://coverage.readthedocs.io/en/6.5.0/index.html)

mezotaken commented:
@salco i'll look into it. I know what it is, but i'm not sure how it will work for this exact case where i'm testing literally everything through api.

salco commented:
> @salco i'll look into it. I know what it is, but i'm not sure how it will work for this exact case where i'm testing literally everything through api. Plus while doing it i realised, that the ideal way would be to run sdwebui in docker container as a service and run tests via command line without server polling from .py file

If it's easy and without issue add it, if you find it's too complicated just post the situation and we (anyone in the community) could add it in another pull-request base on your report. :smile:

Margen67 commented on .github/workflows/run_tests.yaml:
```diff
@@ -0,0 +1,33 @@
+          restore-keys: |
+            ${{ runner.os }}-pip-
+      - name: Run tests
+        run: |
+          python launch.py --tests basic_features --no-half --disable-opt-split-attention --use-cpu all --skip-torch-cuda-test --clip-models-path ./test/test_files/empty.pt
```
```suggestion
        run: python launch.py --tests basic_features --no-half --disable-opt-split-attention --use-cpu all --skip-torch-cuda-test --clip-models-path ./test/test_files/empty.pt
```

Margen67 commented on .github/workflows/run_tests.yaml:
```diff
@@ -0,0 +1,33 @@
+        with:
+          name: stdout-stderr
+          path: |
+            ./test/stdout.txt
+            ./test/stderr.txt
```
```suggestion
            test/stdout.txt
            test/stderr.txt
```

Margen67 commented on .github/workflows/run_tests.yaml:
```diff
@@ -0,0 +1,33 @@
+      - name: Checkout Code
+        uses: actions/checkout@v3
+      - name: Set up Python 3.10
+        uses: actions/setup-python@v3
```
```suggestion
        uses: actions/setup-python@v4
```

mezotaken committed:
Apply suggestions from code review

Use last version of setup-python action
Remove unnecesarry multicommand from run
Remove current directory from artifact paths

Co-authored-by: Margen67 <Margen67@users.noreply.github.com>

mezotaken committed:
Minor fixes

Remove unused test completely
Change job name 
Don't use empty.pt as CLIP weights - it wont work.
Use latest version of actions/cache
